pkg/store/gcs.go

```diff
@@ -0,0 +1,324 @@
+			return err
+		}
+		id, err := ulid.Parse(oi.Prefix[:len(oi.Prefix)-1])
+		if err != nil {
+			continue
+		}
+		if _, ok := s.blocks[id]; ok {
+			continue
+		}
+		level.Info(s.logger).Log("msg", "sync block from GCS", "id", id)
+
+		if err := s.syncBlock(ctx, id); err != nil {
+			level.Error(s.logger).Log("msg", "syncing block failed", "err", err, "block", id.String())
+		}
+	}
+	return nil
+}
+
+// loadFromDisk loads a block with the given ID from the disk cache.
+func (s *GCSStore) loadFromDisk(id ulid.ULID) (*gcsBlock, error) {
+	dir := filepath.Join(s.dir, id.String())
+
+	indexr, err := tsdb.NewFileIndexReader(filepath.Join(dir, "index"))
+	if err != nil {
+		return nil, errors.Wrap(err, "open index reader")
+	}
+	b := &gcsBlock{indexr: indexr, dir: dir}
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	s.blocks[id] = b
+	return b, nil
+}
+
+// syncBlock downloads the index and meta.json file for the given block ID and opens a reader
+// against the block.
+func (s *GCSStore) syncBlock(ctx context.Context, id ulid.ULID) error {
+	bdir := filepath.Join(s.dir, id.String())
+	tmpdir := bdir + ".tmp"
+
+	if err := os.MkdirAll(tmpdir, 0777); err != nil {
+		return errors.Wrap(err, "create temp dir")
+	}
+
+	indexObj := s.bucket.Object(path.Join(id.String(), "index"))
+
+	f, err := os.Create(filepath.Join(tmpdir, "index"))
+	if err != nil {
+		return errors.Wrap(err, "create local index copy")
+	}
+	r, err := indexObj.NewReader(ctx)
+	if err != nil {
+		return errors.Wrap(err, "create index object reader")
+	}
+	_, err = io.Copy(f, r)
+
+	if err := f.Close(); err != nil {
+		level.Warn(s.logger).Log("msg", "close file", "err", err)
+	}
+	if err := r.Close(); err != nil {
+		level.Warn(s.logger).Log("msg", "close object reader", "err", err)
+	}
+	if err != nil {
+		if err := os.RemoveAll(tmpdir); err != nil {
+			level.Warn(s.logger).Log("msg", "cleanup temp dir after failure", "err", err)
+		}
+		return errors.Wrap(err, "copy index file to disk")
+	}
+
+	if err := renameFile(tmpdir, bdir); err != nil {
+		return errors.Wrap(err, "rename block directory")
+	}
+
+	_, err = s.loadFromDisk(id)
+	return errors.Wrap(err, "load synced block")
+}
+
+type gcsBlock struct {
+	dir    string
+	indexr tsdb.IndexReader
+	chunks tsdb.ChunkReader
+}
+
+// Info implements the storepb.StoreServer interface.
+func (s *GCSStore) Info(context.Context, *storepb.InfoRequest) (*storepb.InfoResponse, error) {
+	// Store nodes hold global data and thus have no labels.
+	return &storepb.InfoResponse{}, nil
+}
+
+// Series implements the storepb.StoreServer interface.
+func (s *GCSStore) Series(ctx context.Context, req *storepb.SeriesRequest) (*storepb.SeriesResponse, error) {
+	resp := &storepb.SeriesResponse{}
+
+	matchers, err := translateMatchers(req.Matchers)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	var g errgroup.Group
+	// TODO(fabxc): filter inspected blocks by labels.
+
+	for _, b := range s.blocks {
+		b := b
```

Comment:
?

Reply:
We start a goroutine below. If we don't do this, `b` will be overwritten.
It's the equivalent of 

```go
for _, b := range blocks {
    go func(b Block) {
        // ...
    }(b)
}
```

Just that we don't cal the func ourself with `errgroup`.

Reply:
yeah, standard loop + go routine/defer behavior... I guess you could change the resulting variable name to be bit different to be more explicit? (someone will miss that line while reading code and will look for `b` to check if you pack it into go routine or errgroup)